File: backend/config/slidingWindow.js
const redisClient = require("./redis");
const fs = require("fs");
const path = require("path");

const luaScript = fs.readFileSync(
    path.join(__dirname, "slidingWindow.lua"),
    "utf8"
);


const checkSlidingWindow = async (key, limit, windowSize) => {

    const now = Date.now();

    const requestId = `${now}-${Math.random()}`;

    const result = await redisClient.eval(
        luaScript,
        {
            keys: [key],
            arguments: [
                limit.toString(),
                windowSize.toString(),
                now.toString(),
                requestId
            ]
        }
    );


    console.log("Lua result:", result);

    return {
        allowed: result[0] === 1,
        remaining: Number(result[1]),
        reset: Number(result[2])
    };
};


module.exports = checkSlidingWindow;

File: backend/config/slidingWindow.lua
local key = KEYS[1]

local limit = tonumber(ARGV[1])
local windowSize = tonumber(ARGV[2])
local now = tonumber(ARGV[3])
local requestId = ARGV[4]


local windowStart = now - windowSize


redis.call(
    "ZREMRANGEBYSCORE",
    key,
    0,
    windowStart
)


local count = redis.call(
    "ZCARD",
    key
)


if count >= limit then
    local ttl = redis.call(
        "TTL",
        key
    )

    return {0,0,ttl}
end


redis.call(
    "ZADD",
    key,
    now,
    requestId
)


redis.call(
    "EXPIRE",
    key,
    math.ceil(windowSize / 1000)
)

local remaining = limit - count - 1

local reset = redis.call(
    "TTL",
    key
)


return {1,
remaining,
reset
}